### ISSUE TITLE
Feature request: possibility to pass weight function to A* shortest path search

### ISSUE THREAD

**Issue by esidebo** [networkx/networkx#3774]
Is there a reason the A* path finder `astar.astar_path(G, source, target, heuristic=None, weight='weight')` does not allow for a weight function like is possible for e.g. dijkstra? I'd like to hide specific edges depending on their attributes by setting their weight to `None` like is exemplified under "Notes" at https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.shortest_paths.weighted.single_source_dijkstra.html#networkx.algorithms.shortest_paths.weighted.single_source_dijkstra. 

 I just tried myself to allow for a function by manually changing after line https://github.com/networkx/networkx/blob/b8460be7b055dc18e2fdde3b8739d41f0c0064a3/networkx/algorithms/shortest_paths/astar.py#L114
to instead of calling `w.get(weight, 1)` do:
```
wgt = weight(curnode, neighbor, w)
if wgt is None:
    continue # ignore edge
ncost = dist + wgt
```
I didn't test this thoroughly but seems to work nicely on my small example (i.e. edges with `None` are avoided in the solution). If I'm not missing something, perhaps this could be added?

**Comment by dschult** [comment]
See #3754
We have a number of places where we should be using a weight function (or string) and currently we're only using is as a string. We're definitely fixing this for next version.

**Comment by dschult** [comment]
See also #3797 for a related feature request.
I'm reopening this until fixed along with #3754 and #3797

**Comment by cmkpl** [comment]
I can see from the docs of 2.5rc1.dev_20200420161240, it seems that the function astar_path still only accept string for the parameter weight
